Geometry/Surfaces/QuadSurfaces/cone_class.f90

```diff
@@ -0,0 +1,417 @@
+  !!   a*d^2 + 2b*d + c = 0
+  !!   c = cone surface expression
+  !!   b = (r1 - v1)u1 + (r2 - v2)u2 - t^2(r3 - v3)u3
+  !!   a = u1^2 + u2^2 - t^2 * u3^2
+  !!
+  pure function distance(self, r, u) result(dist)
+    class(cone), intent(in)                 :: self
+    real(defReal), dimension(3), intent(in) :: r
+    real(defReal), dimension(3), intent(in) :: u
+    real(defReal)                           :: dist
+    real(defReal), dimension(3)             :: diff
+    real(defReal), dimension(2)             :: testPoint2
+    real(defReal)                           :: a, b, c, delta, d1, d2, testPoint1, &
+                                               testRadius, maxRadius, distBasis
+
+    ! Vector for the difference between position provided and vertex
+    diff = r - self % vertex
+
+    ! Calculate quadratic components in the cone
+    a = dot_product(u(self % plane), u(self % plane)) - &
+        self % tan2 * u(self % axis)**2
+    b = dot_product(diff(self % plane) , u(self % plane)) - &
+        self % tan2 * diff(self % axis) * u(self % axis)
+    c = dot_product(diff(self % plane), diff(self % plane)) - &
+        self % tan2 * diff(self % axis) ** 2
+
+    ! Calculate delta/4
+    delta = b * b - a * c
+
+    ! Calculate the distances from the cone surface
+    if (a == ZERO) then
+      d1 = - HALF * c / b
+      d2 = d1
+
+    elseif (delta < ZERO) then
+      d1 = INF
+      d2 = d1
+
+    else
+      d1 = - (b + sqrt(delta)) / a
+      d2 = - (b - sqrt(delta)) / a
+
+    end if
+
+    ! If one the distance found is smaller than the surface tolerance, then the particles
+    ! is on a surface: set that distance to infinity
+    if (d1 < self % surfTol()) d1 = INF
+    if (d2 < self % surfTol()) d2 = INF
+
+    ! Keep the smallest positive distance between the two calculated (cone surface)
+    if (d1 >= ZERO .and. d2 >= ZERO) then
+      dist = min(d1, d2)
+    elseif (d1 < ZERO .and. d2 < ZERO) then
+      dist = INF
+    else
+      dist = max(d1, d2)
+    end if
+
+    ! Check that point found would be inside the cone
+    testPoint1 = r(self % axis) + dist * u(self % axis) - self % vertex(self % axis)
+    if (testPoint1 > self % hMax .or. testPoint1 < self % hMin) dist = INF
+
+    ! Calculate the distances from the basis of the cone
+    d1 = (self % hMax - diff(self % axis)) / u(self % axis)
+    d2 = (self % hMin - diff(self % axis)) / u(self % axis)
+
+    ! Verify if the points found would fall inside the cone
+    testPoint2 = r(self % plane) + d1 * u(self % plane) - self % vertex(self % plane)
+    testRadius = dot_product(testPoint2, testPoint2)
+    maxRadius  = self % hMax * self % hMax * self % tan2
+    if (testRadius > maxRadius) d1 = INF
+
+    testPoint2 = r(self % plane) + d2 * u(self % plane) - self % vertex(self % plane)
+    testRadius = dot_product(testPoint2, testPoint2)
+    maxRadius  = self % hMin * self % hMin * self % tan2
+    if (testRadius > maxRadius) d2 = INF
```

Comment:
As an optimisation we can switch to the approach similar to the one  used in the `box`... basically it involves calculating the ranges on the line that fall within a particular body (e.g. inside of the cylinder or space between a pair of planes) and then performing the intersections on the ranges. But is is sadly a bit difficult to explain in the short comment ;-( We can talk about it in person perhaps. 

That being said we can leave the explicit approach if it works well and include the alternative as a TODO in the comment.

Reply:
No that's ok, over the weekend I will have a look at the box and fix this! But I would still need to calculate explicitly whether the particle would intersect the circles? Because in a box or cylinder the two faces are of the same size but this is not true here.

Reply:
Ok, the new distance calculation should be more like box and truncated cylinder now. The only thing I had to add is a check to make sure that the intersection found is on the right half-cone. Otherwise, the segment between two intersections could have been outside of the cone. If an intersection is on the wrong side of the cone, the distance is set to + or - infinity in a way that the intersection 'segment' (or ray) crosses the truncated cone (when here is an actual crossing). This might not be clear, in case let me know!